pitop/pma/ultrasonic_sensor_base.py

```diff
@@ -0,0 +1,347 @@
+from gpiozero import SmoothedInputDevice
+from threading import Thread, Event, Lock
+
+from pitopcommon.logger import PTLogger
+
+from abc import abstractmethod
+from .plate_interface import PlateInterface
+from .common.ultrasonic_registers import (
+    UltrasonicRegisters,
+    UltrasonicRegisterTypes,
+    UltrasonicConfigSettings,
+)
+from pitopcommon.firmware_device import FirmwareDevice
+from pitopcommon.common_ids import FirmwareDeviceID
+from pitop.pma.common.utils import get_pin_for_port
+import atexit
+from sched import scheduler
+import time
+from collections import deque
+import numpy as np
+from pitop.system.peripherals import connected_plate
+
+
+class UltrasonicSensorBase:
+    _max_distance = None
+    threshold = None
+
+    @property
+    def max_distance(self):
+        return self._max_distance
+
+    @max_distance.setter
+    def max_distance(self, value):
+        if value <= 0:
+            raise ValueError('invalid maximum distance (must be positive)')
+        t = self.threshold_distance
+        self._max_distance = value
+        self.threshold_distance = t
+
+    @property
+    def threshold_distance(self):
+        return self.threshold * self.max_distance
+
+    @threshold_distance.setter
+    def threshold_distance(self, value):
+        self.threshold = value / self.max_distance
+
+    @property
+    def distance(self):
+        return self.value * self._max_distance
+
+    @abstractmethod
+    def in_range(self):
+        pass
```

Comment:
```suggestion
        raise NotImplementedError
```

Reply:
If it's being inherited and is required, it should raise an exception when it hasn't been explicitly overridden by the subclass and it is called.

Reply:
(same for all `pass`es)

Reply:
yep!